src/Chainweb/Pact/PactService.hs

```diff
@@ -646,12 +650,28 @@ execLocal cwtx preflight sigVerify rdepth = withDiscardedBatch $ do
     spv <- use psSpvSupport
 
     let rewindHeight
-          | Just d <- rdepth = Just d
+          | Just d <- rdepth = d
           -- when no height is defined, treat
           -- withCheckpointerRewind as withCurrentCheckpointer
           -- (i.e. setting rewind to 0).
-          | otherwise = Just 0
-        rewindHeader = Just $ _tcParentHeader ctx
+          | otherwise = 0
+
+    when ((_height rewindHeight) > _psLocalRewindDepthLimit) $ do
+        throwM $ LocalRewindLimitExceeded (fromIntegral _psLocalRewindDepthLimit) rewindHeight
+
+    let parentBlockHeader = _parentHeader parent
+
+    -- we fail if the requested depth is bigger than the current parent block height
+    -- because we can't go after the genesis block
+    when (fromMaybe 0 rdepth > _blockHeight parentBlockHeader) $ throwM LocalRewindGenesisExceeded
+
+    let ancestorRank = fromIntegral $ _height $ _blockHeight parentBlockHeader - fromMaybe 0 rdepth
+    ancestor <- liftIO $ seekAncestor _psBlockHeaderDb parentBlockHeader ancestorRank
+
+    let rewindHeader
+          | Just a <- ancestor = Just $ ParentHeader a
+          -- when there is no ancestor, use the current parent
```

Comment:
Maybe we want to throw or log an error here if `seekAncestor` returns `Nothing`, I don't see any legal situation where that can happen.

Reply:
Yeah, that behavior is caught by throwing `LocalRewindGenesisExceeded` above, maybe should replace the check with just throwing an exception here.

Reply:
Fixed by throwing the `BlockHeaderLookupFailure` exception because it's indeed an exceptional situation and shouldn't happen.